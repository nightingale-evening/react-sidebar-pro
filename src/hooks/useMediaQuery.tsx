import React from 'react';

export const useMediaQuery = (breakpoint?: string): boolean => {
  const [matches, setMatches] = React.useState(
    breakpoint === 'always' || window.matchMedia(`(max-width: ${breakpoint})`).matches,
  );

  React.useEffect(() => {
    if (breakpoint && breakpoint !== 'always') {
      const media = window.matchMedia(`(max-width: ${breakpoint})`);

      const handleMatch = () => {
        if (media.matches !== matches) {
          setMatches(media.matches);
        }
      };

      handleMatch();

      window.addEventListener('resize', handleMatch);
      return () => window.removeEventListener('resize', handleMatch);
    }
  }, [matches, breakpoint]);

  return matches;
};
