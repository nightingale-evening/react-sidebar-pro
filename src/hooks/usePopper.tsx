import React from 'react';
import { createPopper } from '@popperjs/core';
import { useSidebar } from './useSidebar';

interface PopperOptions {
  level: number;
  buttonRef: React.RefObject<HTMLAnchorElement>;
  contentRef: React.RefObject<HTMLDivElement>;
}

interface PopperResult {
  popperInstance?: ReturnType<typeof createPopper>;
}

export const usePopper = (options: PopperOptions): PopperResult => {
  const [popperInstance, setPopperInstance] = React.useState<ReturnType<typeof createPopper>>();
  const { collapsed, transitionDuration, toggled } = useSidebar();

  const { level, buttonRef, contentRef } = options;

  React.useEffect(() => {
    if (level === 0 && collapsed && contentRef.current && buttonRef.current) {
      const instance = createPopper(buttonRef.current, contentRef.current, {
        placement: 'right',
        strategy: 'fixed',
        modifiers: [
          {
            name: 'offset',
            options: {
              offset: [0, 5],
            },
          },
        ],
      });

      setPopperInstance(instance);
    }
  }, [level, collapsed, contentRef, buttonRef]);

  React.useEffect(() => {
    if (contentRef.current && buttonRef.current) {
      const ro = new ResizeObserver(() => {
        popperInstance?.update();
      });

      ro.observe(contentRef.current);
      ro.observe(buttonRef.current);
    }

    setTimeout(() => {
      popperInstance?.update();
    }, transitionDuration);
  }, [popperInstance, transitionDuration, toggled, contentRef, buttonRef]);

  return { popperInstance };
};
