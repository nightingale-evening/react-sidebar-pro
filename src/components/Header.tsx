import React from 'react';
import styled from 'styled-components';
import classnames from 'classnames';
import { useSidebar } from '../hooks/useSidebar';

export interface HeaderProps extends React.HTMLAttributes<HTMLHeadElement> {
  height?: string;
  fixed?: boolean;
}

interface StyledHeaderProps extends HeaderProps {
  collapsedSidebar?: boolean;
  brokenSidebar?: boolean;
  fixedSidebar?: boolean;
  sidebarWidth?: string;
  sidebarCollapsedWidth?: string;
  transitionDuration: number;
  mounted: boolean;
}

const StyledHeader = styled.header<StyledHeaderProps>`
  height: ${({ height }) => height};
  min-height: ${({ height }) => height};
  position: relative;
  width: 100%;
  ${({ mounted, transitionDuration }) =>
    mounted ? `transition: width ${transitionDuration}s;` : ''}

  ${({ fixed, fixedSidebar, collapsedSidebar, sidebarWidth, sidebarCollapsedWidth, height }) =>
    fixed
      ? `
      position: fixed;
      z-index: 2;
        ~ .layout,
        ~ .content {
          margin-top: ${height};
        }

        ${
          fixedSidebar
            ? `width:calc(100% - ${collapsedSidebar ? sidebarCollapsedWidth : sidebarWidth})`
            : ''
        }
        `
      : ''};

  ${({ brokenSidebar }) => (brokenSidebar ? 'width: 100% !important;transition: none;' : '')}
`;

export const Header: React.FC<HeaderProps> = ({
  height = '64px',
  fixed = false,
  className,
  children,
  ...rest
}) => {
  const [mounted, setMounted] = React.useState(false);

  const {
    collapsed: collapsedSidebar,
    fixed: fixedSidebar,
    width: sidebarWidth,
    collapsedWidth: sidebarCollapsedWidth,
    broken: brokenSidebar,
    transitionDuration,
  } = useSidebar();

  React.useLayoutEffect(() => {
    setMounted(true);
  }, []);

  return (
    <StyledHeader
      data-testid="header-test-id"
      mounted={mounted}
      height={height}
      fixed={fixed}
      collapsedSidebar={collapsedSidebar}
      brokenSidebar={brokenSidebar}
      fixedSidebar={fixedSidebar}
      sidebarWidth={sidebarWidth}
      sidebarCollapsedWidth={sidebarCollapsedWidth}
      transitionDuration={(transitionDuration ?? 300) / 1000}
      className={classnames('header', className)}
      {...rest}
    >
      {children}
    </StyledHeader>
  );
};
