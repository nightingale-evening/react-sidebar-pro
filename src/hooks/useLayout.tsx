import React from 'react';
import { LayoutContext, LayoutContextProps } from '../components/layoutContext';

export const useLayout = (): LayoutContextProps => {
  const context = React.useContext(LayoutContext);
  if (context === undefined) {
    //TODO: set better error message
    throw new Error('ProSidebarProvider is required!');
  }
  return context;
};
