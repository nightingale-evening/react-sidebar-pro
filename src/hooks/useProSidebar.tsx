import { useLayout } from './useLayout';
import { useSidebar } from './useSidebar';

interface ProSidebarFunction {
  /**
   * a function that enables you to update the sidebar's collapsed status
   */
  collapseSidebar: (collapsed?: boolean) => void;

  /**
   * a function that enables you to update the sidebar's toggled status
   */
  toggleSidebar: (toggled?: boolean) => void;

  broken: boolean;

  collapsed: boolean;

  toggled: boolean;

  rtl: boolean;
}

export const useProSidebar = (): ProSidebarFunction => {
  const { updateSidebarState, collapsed, toggled, broken } = useSidebar();
  const { rtl } = useLayout();

  const collapseSidebar = (value?: boolean) => {
    updateSidebarState({ collapsed: value ?? !collapsed });
  };

  const toggleSidebar = (value?: boolean) => {
    updateSidebarState({ toggled: value ?? !toggled });
  };

  return {
    collapseSidebar,
    toggleSidebar,
    collapsed: !!collapsed,
    broken: !!broken,
    toggled: !!toggled,
    rtl: !!rtl,
  };
};
