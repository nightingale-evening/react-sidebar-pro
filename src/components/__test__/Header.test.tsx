import React from 'react';
import { customRender, screen } from '../../utils/testUtils';
import { Header } from '../Header';
import * as hooks from '../../hooks/useSidebar';

describe('Header', () => {
  it('basic snapshot ', () => {
    const { container } = customRender(<Header>Header</Header>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should initialize Header correctly', async () => {
    customRender(<Header>Header</Header>);
    const headerElem = screen.getByText('Header');

    expect(headerElem).toBeInTheDocument();
    expect(headerElem).toHaveClass('header');
    expect(headerElem).toHaveStyle({
      height: '64px',
      'min-height': '64px',
    });
  });

  it('should set the height to 50px', async () => {
    customRender(<Header height="50px">Footer</Header>);
    const headerElem = screen.getByTestId('header-test-id');

    expect(headerElem).toHaveStyle({
      height: '50px',
      'min-height': '50px',
    });
  });

  it('should set have position:fixed when fixed prop is true', async () => {
    customRender(<Header fixed>Footer</Header>);
    const headerElem = screen.getByTestId('header-test-id');

    expect(headerElem).toHaveStyle({
      position: 'fixed',
    });
  });

  it('should have a width of 100% - sidebarWidth when header and sidebar fixed', () => {
    jest.spyOn(hooks, 'useSidebar').mockImplementation(() => ({
      updateSidebarState: jest.fn(),
      collapsed: false,
      fixed: true,
      width: '250px',
      collapsedWidth: '80px',
      broken: false,
      toggled: false,
      transitionDuration: 300,
    }));

    customRender(<Header fixed>Header</Header>);

    const headerElem = screen.getByTestId('header-test-id');

    expect(headerElem).toHaveStyle({
      width: 'calc(100% - 250px)',
    });
  });

  it('should have a width of 100% - sidebarCollapsedWidth when header and sidebar fixed and collapsed', () => {
    jest.spyOn(hooks, 'useSidebar').mockImplementation(() => ({
      updateSidebarState: jest.fn(),
      collapsed: true,
      fixed: true,
      width: '250px',
      collapsedWidth: '80px',
      broken: false,
      toggled: false,
      transitionDuration: 300,
    }));

    customRender(<Header fixed>Header</Header>);

    const headerElem = screen.getByTestId('header-test-id');

    expect(headerElem).toHaveStyle({
      width: 'calc(100% - 80px)',
    });
  });

  it('should have a width of 100% and transition none when sidebar is broken', () => {
    jest.spyOn(hooks, 'useSidebar').mockImplementation(() => ({
      updateSidebarState: jest.fn(),
      collapsed: false,
      fixed: false,
      width: '250px',
      collapsedWidth: '80px',
      broken: true,
      toggled: false,
      transitionDuration: 300,
    }));

    customRender(<Header fixed>Header</Header>);

    const headerElem = screen.getByTestId('header-test-id');

    expect(headerElem).toHaveStyle({
      width: '100%',
      transition: 'none',
    });
  });
});
