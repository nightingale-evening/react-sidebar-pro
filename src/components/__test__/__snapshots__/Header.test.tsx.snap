// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header basic snapshot  1`] = `
<header
  class="sc-bdvvtL kGHiwU header"
  data-testid="header-test-id"
  height="64px"
>
  Header
</header>
`;
